### ISSUE TITLE
IME Support

### ISSUE THREAD

**Issue by buckle2000** [openfl/openfl#1098]
please support ime as it is important to i18n.
currently ime is always on on flash target, but off on others.
there is no way to control the status of ime, as I know.

**Comment by larsiusprime** [comment]
I need this too as I will be soon adding Japanese and Korean localization to Defender's Quest.

If anyone has some good ideas for 3rd party libraries or native hooks that could enable this, let me know and I can start work on an implementation, if necessary.

**Comment by larsiusprime** [comment]
Did some quick digging, it looks like SDL might be IME-aware. I think OpenFL might need to implement some way to render the IME interface in-app, but we could base it off how it's implemented in Flash...

https://wiki.libsdl.org/Tutorials/TextInput

I'll keep digging. Does anyone know how well IME works through raw SDL? Are there any pitfalls or gotchas, and/or is there a better cross-platform solution we should hook into instead?

**Comment by larsiusprime** [comment]
Okay got some good answers from @mauvecow -- looks like SDL is the way to go, and it seems if you properly interface with the API the operating system should also provide the correct text overlay.

**Comment by larsiusprime** [comment]
Okay so got some extra feedback from sikthehedgehog on twitter re:SDL, posting here as a reference:

> OK I think we discussed all the important points (and probably most of this doesn't apply if you're using a recent version) but anyway, putting allt his together for future reference:
> 
> - 2.0.3 didn't support IME on Linux in any form
> - 2.0.4 does support it (2.0.4-9174 or later is guaranteed to)
> 
> Early implementations of the IME support break on anything that isn't Debian/Ubuntu due to not checking a parameter with dBus. You probably won't hit this, but if you do, then update SDL, it's not a binary compatibility issue (saying this as somebody who had bingcc set up).
>
> Gnome provides a built-in field that lets you see what you're typing even if the program doesn't render this explicitly. Sadly, it only works in windowed mode, since the field isn't topmost (non-SDL programs suffer from this too, only workaround is to create the window as windowed then make it fullscreen). Given you don't have this feature on Windows anyway you may as well not rely on it.
>
> On that note, there's a hint called "SDL_IME_INTERNAL_EDITING" (aka the constant SDL_HINT_IME_INTERNAL_EDITING). It controls the aforementioned feature. When it's 0 you have to render it yourself, when it's 1 then Gnome will show the field instead. It's important since if you do it yourself you really want to disable this :P
>
> I think that covers everything. Have fun :O)

**Comment by larsiusprime** [comment]
So right now it looks like the process is:

1. Listen for / respond to the right text input events in SDL
2. Decide whether to render the IME composition characters in-app or leave it to the OS overlay

From the looks of it the answer to 2. might be "render in-app" but I'll need to investigate.

**Comment by larsiusprime** [comment]
Further tips:

> Also damn I forgot about one more quirk, but it isn't specific to Linux (it's a limitation of SDL itself). When typing in cyrillic, writing Z types in З. Following it by H should replace it with Ж, but since SDL has no way to tell you to do that instead what you get is ЗЖ (i.e. both characters). As far as I know there's no way to work around it short of having the program explicitly handle it.
>
> I don't know what other issues like this exist out there, but you have been warned. At least if you want to play nice with Russian players :O)

**Comment by sikthehedgehog** [comment]
If you want to support anything that isn't Linux (even Windows!) then "in-app" is pretty much the only option (the OS overlay in Linux exists mainly to make it work even in programs that don't explicitly support it - Code::Blocks comes to mind, from experience). Sadly unless you're an UI toolkit this means every program will have to provide its own means to do it (as the program draws itself).

On the other hand the overlay makes it easier to test incomplete IME code if you're on Linux. Or you could be faking the overlay on your own on behalf of the programs as a quick alternative.

**Comment by jgranick** [comment]
Hmm, I thought an issue like the above Cyrillic input issue was managed by SDL "text editing" vs SDL "text input" events, maybe this is not the case

**Comment by larsiusprime** [comment]
Well, the advantage of having to render the composition ourselves means that we can ensure that it's reasonably similar cross-platform, I guess.

**Comment by jgranick** [comment]
Does anyone know how to trigger it in Flash Player? I've never seen IME input before

**Comment by sikthehedgehog** [comment]
Text editing is used for things like Japanese input where it takes multiple keystrokes for just about anything, so you're entering partial text and then have to press Enter. But Cyrillic doesn't work that way, for the most part is "keystroke → character", the ZH override is the exception to the rule (and works by replacing the first character if you type the second one right after it).

**Comment by larsiusprime** [comment]
@jgranick -- you can see it in action in the Steam version of Defender's Quest.

**Comment by larsiusprime** [comment]
@jgranick also is air only i believe:
http://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/system/IME.html

**Comment by larsiusprime** [comment]
@jgranick I also wonder if this will resolve some other issues we've had with unicode text input

**Comment by buckle2000** [comment]
@larsiusprime maybe you can use target specific API, [ImmAssociateContext](https://msdn.microsoft.com/en-us/library/windows/desktop/dd318171.aspx) on Windows for example

**Comment by larsiusprime** [comment]
@buckle2000 we generally try to avoid target-specific solutions because it kind of explodes the maintenance burden, but if it's a simple enough call we'll look into it. We'll know more when we dig into the implementation.

**Comment by buckle2000** [comment]
@larsiusprime I have tried it in c++, see [here](https://gist.github.com/buckle2000/6b6000e3c6fcce8e893906d27d05a631)

**Comment by jgranick** [comment]
Tracking here: https://trello.com/c/4xoodFKz/47-add-support-for-ime-text-input

**Comment by apptaro** [comment]
@jgranick so what's the status for TextField's IME support? OpenFL is not usable without it for Japanese users.

**Comment by apptaro** [comment]
FYI, IME doesn't work for html5 target, either, but I found something interesting; this library draws input field on canvas and IME does work magically for Japanese.
https://github.com/goldfire/CanvasInput

**Comment by jgranick** [comment]
Certainly, it might be possible to compare notes with the CanvasInput library, and to see how it supports IME. I think these would be the relevant places to compare:

https://github.com/goldfire/CanvasInput/blob/master/CanvasInput.js
https://github.com/openfl/lime/blob/develop/lime/_backend/html5/HTML5Window.hx#L772

Perhaps one possibility is that I believe we clear our input (after we dispatch the events), rather than preserving it all in the hidden field

**Comment by minemaz** [comment]
This implementation is probably helpful.

https://github.com/apptaro/PIXI.Input2
https://apptaro.github.io/PIXI.Input2/test.html

**Comment by arloan** [comment]
Well, this problem's still not solved for a decade. Not to talk about IME input, simply paste some CJK text into the textfield, nothing will be shown except the caret position moved. If u move the caret with arrow keys, the app crashes immediately.
